File: tri_predictr.js
const _ = require('lodash');
const ngrams = { _count: 0 };
const deep = 2;

const updateGram = (base, word) => {
	const n = base._count;
	if (!_.has(base, word)) {
		base[word] = { _probability: 0, _count: 0 };
	}
	for (let key of Object.keys(base)) {
		if (_.includes(['_probability', '_count'], key)) continue;
		const tmp = base[key]._probability;
		base[key]._probability = (tmp * n + (key === word ? 1 : 0)) / (n + 1);
	}
	base._count++;
};

const iterate = (base, sentence, index, number) => {
	if (number > index) number = index;
	while (number) {
		base = base[sentence[index - number]];
		number--;
	}
	return (base);
};

const deepLearn = (base, sentence, index, number) => {
	if (number > index) number = index;
	const tmp = iterate(base, sentence, index, number);
	updateGram(tmp, sentence[index]);
	if (number) deepLearn(base, sentence, index, number - 1);
};

const learn = str => {
	const sentence = str.split(' ');
	_.times(sentence.length, index => {
		deepLearn(ngrams, sentence, index, deep);
	});
};

const predict = (sentence) => {
	const index = sentence.length - 1;
	let res = ['', 0];
	const base = iterate(ngrams, sentence, index, deep - 1)[sentence[index]];
	for (let key of Object.keys(base)) {
		if (!_.includes(['_probability', '_count'], key) && base[key]._probability > res[1]) {
			res = [key, base[key]._probability];
		}
	}
	return res[0];
};

(() => {
	const test = 'a b c d';
	const predTo = 'a b c';
	const test2 = 'a b c e';
	const test3 = 'a b c e';
	const predTo2 = 'a b c';

	learn(test);
	if ('d' === predict(predTo.split(' '))) {
		console.log('test succesful');
	}
	learn(test2);
	learn(test3);
	if ('e' === predict(predTo.split(' '))) {
		console.log('test succesful');
	}
})();

File: tri_predictr.py
# Les trigrams seront enregistres en memoire de la maniere suivante
# 1 gram { 'w': 2-gram, '_n': nb_bigram, '_p': proba }
#    ---> 2 gram { 'w': 3-gram, '_n': nb_trigram, '_p': proba }
#            ---> 3 gram { '_p': proba }

# from flask import Flask
import json

# app = Flask(__name__)

class markov():
    """ markov is the class engine for word prediction.

    Attribute:
        ngrams: tree of the ngrams
        deep: the depth of the tree
    """

    def __init__(self, deep = 0):
         self.ngrams = { '_n': 0 } # contient les trigrams
         self.deep = deep

    def __update_gram(self, base, word):
        n = base['_n']
        if word not in base:
            base[word] = { '_p': 0, '_n': 0 }
        for x in base:
            if x == '_p' or x == '_n':
                continue
            tmp = base[x]['_p']
            base[x]['_p'] = (tmp * n + (1 if x == word else 0)) / (n + 1)
        base['_n'] += 1

    def __iterate(self, base, sentence, i, n):
        while n > i:
            n -= 1
        while n > 0:
            base = base[sentence[i - n]]
            n -= 1
        return (base)

    def __deep_learn(self, base, sentence, i, n):
        tmp_base = self.__iterate(base, sentence, i, n)
        self.__update_gram(tmp_base, sentence[i])
        if n > 0:
            self.__deep_learn(base, sentence, i, n - 1)

    def learn(self, s):
        sentence = s.split(' ')
        for i in range(len(sentence)):
            self.__deep_learn(self.ngrams, sentence, i, self.deep)

    def predict(self, sentence, i):
        res = ('', 0)
        base = self.__iterate(self.ngrams, sentence, i, self.deep - 1)
        print(base)
        base = base[sentence[i]]
        tuple_proba = lambda x: (x, base[x]['_p'])
        res = sorted(map(tuple_proba, base.keys() ^ {'_p', '_n'}), key = lambda x: x[1], reverse = True)
        return res[:5]
        for x in base:
            if x not in ['_p', '_n'] and base[x]['_p'] > res[1]:
                res = (x, base[x]['_p'])
        return res[:5]

    def show(self):
        print(self.ngrams)

# @app.route('/')
# def preditr():
#     markov = predictr.__markov


if __name__ == '__main__':
    IA = markov(2)
#    predictr.__markov = IA
    while True:
        line = input().strip()
        if line.strip() == 'exit':
            print('done')
            exit()
        if line.strip() == 'show':
            print(IA.show())
            continue
        IA.learn(line)
        tab = line.split(' ')
        print(IA.predict(tab, len(tab) - 1))
